### ISSUE TITLE
[Feature Request] Consider Upstreaming Optional MoE Routing and Detector-Aware PEFT Support

### ISSUE THREAD

**Issue by isLinXu** [ultralytics/ultralytics#26110]
### Search before asking

- [x] I have searched the Ultralytics [issues](https://github.com/ultralytics/ultralytics/issues) and found no similar feature requests.

### Search before asking

- [x] I have searched the existing Ultralytics issues and found no equivalent feature request.

### Description

Hi @glenn-jocher and the Ultralytics team,

I am Xu Lin, the first author of the YOLO-Master and YOLO-PEFT papers.

I would like to ask for maintainer guidance before preparing any pull request. We have developed two optional research features on top of the Ultralytics codebase:

1. **YOLO-Master: Efficient Sparse Mixture-of-Experts for real-time detection**

   YOLO-Master introduces ES-MoE blocks and lightweight dynamic routing for instance-conditional computation. The implementation includes YAML-defined MoE model configurations, top-k expert routing, auxiliary load-balancing/stability losses, and expert utilization diagnostics.

2. **YOLO-PEFT: Structure-aware Parameter-Efficient Fine-Tuning for YOLO**

   YOLO-PEFT treats adapter placement as an auditable planning problem. It analyzes the detector graph, validates candidate target modules, allocates adapter budgets, rejects unsafe placements before training, and supports train/save/merge/export workflows.

This issue is not requesting that the entire YOLO-Master repository be merged into Ultralytics as-is. Instead, I would like to confirm whether either of these capabilities is relevant to the Ultralytics roadmap and whether the team would be open to a focused upstream implementation.

### Proposed upstream scope

| Track | Suggested initial upstream scope | Not required in the first PR |
|---|---|---|
| Optional MoE routing | A generic MoE/router module, YAML configuration support, optional auxiliary-loss hooks, unit tests, and clear export/backend behavior | All experimental variants, MoE pruning, CW-NMS, Sparse SAHI, and backend-specific optimizations |
| Detector-aware PEFT | Optional adapter registration, target-module validation, LoRA/RS-LoRA support, adapter save/merge/export, and documentation | Every adapter algorithm, the full research planner, and calibration policies for unsupported architectures |

The two tracks can be implemented and reviewed independently.

### Why this may be useful

- MoE routing can allocate more capacity to difficult or crowded scenes while keeping the default dense YOLO path unchanged.
- PEFT can reduce the cost of adapting YOLO models to new domains, datasets, or industrial scenarios.
- Merged adapters can be exported as ordinary YOLO models without inference-time adapter overhead.
- Both features can remain opt-in and should not change existing model behavior or default configurations.

### Compatibility and maintenance constraints

We would follow the following principles:

- Existing Ultralytics models and training commands remain backward compatible.
- Additional PEFT dependencies remain optional.
- MoE deployment behavior is documented explicitly for each backend; no assumption is made that every runtime provides actual sparse-computation speedups.
- New functionality is covered by unit tests, training/validation/inference tests, export checks, and baseline comparisons.
- The first contribution would be narrowly scoped and aligned with the Ultralytics API and coding style.

### Questions for the maintainers

1. Would optional MoE or conditional-routing support be considered within the Ultralytics roadmap?
2. Would generic detector-aware PEFT/LoRA support be considered within the roadmap?
3. If these features are potentially acceptable, would you prefer two separate issues and pull requests?
4. What would be the minimum acceptable API, supported model families, backend coverage, and benchmark requirements?
5. If these features are not suitable for the core repository, would an official extension package or integration guide be preferable?

For reference:

- YOLO-Master: https://github.com/Tencent/YOLO-Master
- YOLO-Master paper: https://arxiv.org/abs/2512.23273
- YOLO-PEFT paper: https://arxiv.org/abs/2608.07051

The published results are research results, and we would be happy to reproduce the relevant benchmarks under the Ultralytics evaluation and deployment protocol before submitting a PR.

### Use case

The main use cases are:

- real-time detection under changing scene complexity;
- deployment on compute-constrained edge devices;
- rapid domain adaptation without storing a complete fine-tuned model for every domain;
- reproducible research on efficient YOLO architectures and adaptation.

### Additional

The YOLO-Master repository is released under AGPL-3.0. I am willing to follow the Ultralytics contribution guidelines, CLA requirements, testing requirements, and any maintainer-defined implementation scope.

### Are you willing to submit a PR?

Yes. I am willing to prepare and maintain a focused pull request after the scope and acceptance criteria are approved by the maintainers.

### Are you willing to submit a PR?

- [ ] Yes I'd like to help by submitting a PR!

**Comment by UltralyticsAssistant** [comment]
👋 Hello @isLinXu, thank you for the detailed proposal and for offering to prepare a focused contribution. Optional MoE routing and detector-aware PEFT/LoRA are substantial object-detection features, so the maintainers will need to review their roadmap fit, API scope, supported model families, export/backend behavior, and benchmark requirements before recommending an implementation path. Keeping the two tracks independent, opt-in, and backward compatible is helpful.

This is an automated response, and an Ultralytics engineer will assist soon with guidance on whether these capabilities belong in the core repository or would be better maintained as an extension or integration. In the meantime, please review the [Docs](https://docs.ultralytics.com) and contribution expectations while awaiting maintainer feedback.

**Comment by Y-T-G** [comment]
Do you have benchmarks of improvements with these techniques?

Does PEFT work for CNN based YOLO models?

**Comment by isLinXu** [comment]
Thanks for the questions. Yes, we have benchmark results for both MoE and PEFT. However, we consider these results to be research-stage evidence rather than universal guarantees.

### 1. Benchmarks for MoE

The most direct comparison in our released benchmark table is between YOLOv13-N and YOLO-Master-N:

<img width="797" height="851" alt="Image" src="https://github.com/user-attachments/assets/ccf7f83d-3128-4106-b04b-300b3facb3da" />

| Benchmark | YOLOv13-N | YOLO-Master-N | Improvement |
|---|---:|---:|---:|
| COCO mAP | 41.6 | 42.4 | +0.8 |
| PASCAL VOC mAP50 | 60.7 | 62.1 | +1.4 |
| VisDrone mAP | 17.5 | 19.6 | +2.1 |
| KITTI mAP50 | 67.7 | 69.2 | +1.5 |
| SKU-110K mAP50 | 57.5 | 58.2 | +0.7 |
| TensorRT latency | 1.97 ms | 1.62 ms | 17.8% lower |

The released EsMoE model zoo also reports the following TensorRT results on an RTX 4090:

| Model | Parameters | mAP50-95 | TensorRT FPS |
|---|---:|---:|---:|
| YOLO-Master-EsMoE-N | 2.68M | 0.427 | 640.18 |
| YOLO-Master-EsMoE-S | 9.69M | 0.489 | 423.87 |
| YOLO-Master-EsMoE-M | 34.88M | 0.530 | 243.79 |

These model-zoo numbers should be interpreted as released engineering benchmarks. We would be happy to reproduce the results under the Ultralytics evaluation and deployment protocol before preparing an upstream PR.

### 2. Does PEFT work for CNN-based YOLO models?

<img width="2048" height="1254" alt="Image" src="https://github.com/user-attachments/assets/12109a61-1916-4f66-938b-6831f5ae2872" />

Yes, within the evaluated scope.

We evaluated planner-selected RS-LoRA on YOLO11s and YOLO12s using the official VOC07+12 trainval-to-VOC07 test protocol:

| Model | Full-SFT | Planner-selected RS-LoRA | Difference |
|---|---:|---:|---:|
| YOLO11s | 0.6428 | 0.7138 | +0.0710 |
| YOLO12s | 0.6662 | 0.7307 | +0.0645 |

A controlled YOLO11 audit also showed:

- Peak training memory: 28.57 GB → 16.03 GB, a 43.9% reduction.
- Training time: 1.72× longer than Full-SFT.

The important point is that PEFT is applicable to CNN-based YOLO models, but it is not architecture-agnostic. We do not blindly insert LoRA into every convolution, depthwise/grouped convolution, DFL component, or detection-head module.

Instead, YOLO-PEFT performs structure-aware target selection based on operator validity, detector semantics, graph interfaces, and deployment constraints.

This distinction is important because some architectures and placements are unstable. For example, on RT-DETR-L, all seven evaluated LoRA-family configurations crossed the predefined catastrophic threshold. In that case, the planner returns `Refuse` and falls back to Full-SFT instead of silently producing an unreliable adapter.

Therefore, our conclusion is:

> PEFT does work for CNN-based YOLO models, especially when the adapter placement is selected in a structure-aware manner. However, it should not be advertised as a universally safe drop-in method for every layer or detector architecture.

References:

- YOLO-Master: https://github.com/Tencent/YOLO-Master
- YOLO-Master paper: https://arxiv.org/abs/2512.23273
- YOLO-PEFT paper: https://arxiv.org/abs/2608.07051

We would be happy to provide the training configurations, evaluation scripts, logs, and additional CNN-only ablations requested by the Ultralytics team.

**Comment by isLinXu** [comment]
wandb：https://wandb.ai/linxu/YOLO-Master-Exp-v2607?nw=nwuserlinxu

**Comment by Y-T-G** [comment]
Can PEFT allow users to add classes to a pretrained YOLO model without forgetting old classes?

**Comment by isLinXu** [comment]
Thanks for the question. 
PEFT can help with incremental class learning, but it cannot guarantee no forgetting by itself. YOLO-PEFT freezes most pretrained weights and trains lightweight adapters, which reduces direct interference. However, adding new classes still requires detection-head expansion and an anti-forgetting strategy such as old-data replay or knowledge distillation. The current YOLO-PEFT paper evaluates fixed-vocabulary adaptation and does not yet include a class-incremental benchmark. 
Therefore, we do not claim that it supports adding new classes without forgetting old classes out of the box. This is a natural next extension, which we would be happy to benchmark using old-class mAP, new-class mAP, all-class mAP, and forgetting rate.

**Comment by Y-T-G** [comment]
PEFT would be useful to users if it can aid incremental training. We would also like a fine-tuning test. But next week after YOLO Vision. We have some announcements that would be relevant.

MoE would be useful if it can aid smaller YOLO models and help improve the accuracy on them without inference overhead.

**Comment by isLinXu** [comment]
Thank you for the valuable feedback. I will be attending YOLO Vision in person on September 13, and I’m looking forward to seeing your latest products and announcements. Congratulations, and see you there!
